{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assembly with the Canu pipeline\n",
    "\n",
    "[Canu](https://github.com/marbl/canu) is a popular assembler based on Celera Assembler and built specifically to work with ONT reads. It consists on a 4-step pipeline that generates a 'draft assembly' without reference. In order to get better results, Canu is often used with tools that improves it result. In this notebook, we will build a popular Canu pipeline using Canu + Racon + Pilon.\n",
    "\n",
    "Canu works with either FASTA or FASTQ files (compressed and uncompressed), but FASTQ format is needed to run the next steps and complete the full pipeline. Help page can be shown with \"canu -h\" command. These are the parameters needed for running Canu with our data:\n",
    "\n",
    "- -p and -d: Assembly files prefix and output directory. Both parameters can be the same and output directory doesn't have to exist before execution.\n",
    "- genomeSize: The estimated genome size. In our case, 2.1 mbp so we write '2.1m'. We can put letter g for gbp or k for kbp as well.\n",
    "- -nanopore-raw: The path to our reads in FASTQ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "canu -p agalactiae \\\n",
    "     -d data/agalactiae/canu_output \\\n",
    "     genomeSize=4.6m \\\n",
    "     useGrid=false \\\n",
    "     minReadLength=50 \\\n",
    "     minOverlapLength=50 \\\n",
    "     -nanopore-raw data/agalactiae/merged-output.fastq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Racon\n",
    "\n",
    "[Racon](https://github.com/isovic/racon) is a consensus module to correct raw contigs generated by rapid assembly methods which do not include a consensus step. We will take our Canu result in FASTA and the raw reads in FASTQ and generate a new FASTA contig file.\n",
    "\n",
    "Before running Racon, we must align the Canu output to the raw reads file and take the overlaps file in PAF as a parameter for Racon command. This can be done using for example minimap, a rapid aligner for ONT reads.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "minimap data/agalactiae/canu_output/agalactiae.contigs.fasta \\\n",
    "        data/agalactiae/merged-output.fastq \\\n",
    "        > agalactaiae.paf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Racon parameters:\n",
    "- -t: Number of threads\n",
    "- Raw reads (FASTQ).\n",
    "- Overlaps in (PAF).\n",
    "- Canu output (FASTA)\n",
    "- Racon output file name (FASTA)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "racon -t 48 \\\n",
    "    data/agalactiae/output/merged-output.fastq \\ \n",
    "    agalactiae.paf \\ \n",
    "    data/agalactiae/canu/agalactiae.contigs.fasta \\ \n",
    "    data/agalactiae_racon.contigs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pilon (Requires Illumina reads)\n",
    "\n",
    "[Pilon](https:github.com/broadinstitute/pilon) is a tool that can be used to improve draft assembly as our case and find variation among strains. It Requires as input a FASTA file and a BAM file of reads aligned to the input FASTA file. At this point, we have the racon contigs file and our reads file (ONT and Illumina reads). We will use the racon contigs file and a BAM file produced by de alignment of ILlumina reads against that contigs file to generate our final result with Pilon.\n",
    "\n",
    "In first place, we need to align our Illumina reads against the racon contigs file. We are using bwa-mem so first we have to index the reference file (racon contigs) and then perform the alignment of the Illumina reads. In order to have the BAM file required by Pilon, we are using samtools to convert bwa output in FASTA to the BAM format.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bwa index data/agalactaie_racon.contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Illumina data not included in repository\n",
    "bwa mem -t 48 \\\n",
    "        data/agalactaie_racon.contigs.fasta  \\\n",
    "        data/Data_Ilumina/Raw/WGS_bacterialIsolates_MiSeq_training-33608589/Sagalactiae_HRC-41106565/Sagal_S5_L001_R1_001.fastq.gz \\\n",
    "        data/Data_Ilumina/Raw/WGS_bacterialIsolates_MiSeq_training-33608589/Sagalactiae_HRC-41106565/Sagal_S5_L001_R2_001.fastq.gz \\\n",
    "        | samtools view -S -b -u - | samtools sort - data/bwa_aligned_reads\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally we index the BAM file an run the Pilon jar:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "samtools index data/bwa_aligned_reads.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pilon arguments:\n",
    "- --threads: Number of threads\n",
    "- --genome: The FASTA input file (racon contigs)\n",
    "- --bam: BAM file (generated by bwa and samtools)\n",
    "- --outdir and --output: Output directory and filename "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "java -Xmx128g -XX:+UseConcMarkSweepGC \\ \n",
    "      -XX:-UseGCOverheadLimit \\ \n",
    "      -jar /home/jovyan/software/pilon/pilon-1.22.jar \\ \n",
    "      --threads 2 \\ \n",
    "      --genome data/agalactaie_racon.contigs.fasta \\ \n",
    "      --bam data/bwa_aligned_reads.bam \\ \n",
    "      --outdir data/agalactiae/pilon_output \\ \n",
    "      --output pilon.contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
