{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assembly with the Canu pipeline\n",
    "\n",
    "[Canu](https://github.com/marbl/canu) is a popular assembler based on Celera Assembler and built specifically to work with ONT reads. It consists on a 4-step pipeline that generates a 'draft assembly' without reference. In order to get better results, Canu is often used with tools that improves it result. In this notebook, we will build a popular Canu pipeline using Canu + Racon + Pilon.\n",
    "\n",
    "Canu works with either FASTA or FASTQ files (compressed and uncompressed), but FASTQ format is needed to run the next steps and complete the full pipeline. Help page can be shown with \"canu -h\" command. These are the parameters needed for running Canu with our data:\n",
    "\n",
    "- -p and -d: Assembly files prefix and output directory. Both parameters can be the same and output directory doesn't have to exist before execution.\n",
    "- genomeSize: The estimated genome size. In our case, 2.1 mbp so we write '2.1m'. We can put letter g for gbp or k for kbp as well.\n",
    "- -nanopore-raw: The path to our reads in FASTQ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Canu snapshot v1.7 +137 changes (r8829 73d5caa1b1087b65f7853ecbebc1bb1dcbd1bc14)\n",
      "--\n",
      "-- CITATIONS\n",
      "--\n",
      "-- Koren S, Walenz BP, Berlin K, Miller JR, Phillippy AM.\n",
      "-- Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation.\n",
      "-- Genome Res. 2017 May;27(5):722-736.\n",
      "-- http://doi.org/10.1101/gr.215087.116\n",
      "-- \n",
      "-- Read and contig alignments during correction, consensus and GFA building use:\n",
      "--   Šošic M, Šikic M.\n",
      "--   Edlib: a C/C ++ library for fast, exact sequence alignment using edit distance.\n",
      "--   Bioinformatics. 2017 May 1;33(9):1394-1395.\n",
      "--   http://doi.org/10.1093/bioinformatics/btw753\n",
      "-- \n",
      "-- Overlaps are generated using:\n",
      "--   Berlin K, et al.\n",
      "--   Assembling large genomes with single-molecule sequencing and locality-sensitive hashing.\n",
      "--   Nat Biotechnol. 2015 Jun;33(6):623-30.\n",
      "--   http://doi.org/10.1038/nbt.3238\n",
      "-- \n",
      "--   Myers EW, et al.\n",
      "--   A Whole-Genome Assembly of Drosophila.\n",
      "--   Science. 2000 Mar 24;287(5461):2196-204.\n",
      "--   http://doi.org/10.1126/science.287.5461.2196\n",
      "-- \n",
      "-- Corrected read consensus sequences are generated using an algorithm derived from FALCON-sense:\n",
      "--   Chin CS, et al.\n",
      "--   Phased diploid genome assembly with single-molecule real-time sequencing.\n",
      "--   Nat Methods. 2016 Dec;13(12):1050-1054.\n",
      "--   http://doi.org/10.1038/nmeth.4035\n",
      "-- \n",
      "-- Contig consensus sequences are generated using an algorithm derived from pbdagcon:\n",
      "--   Chin CS, et al.\n",
      "--   Nonhybrid, finished microbial genome assemblies from long-read SMRT sequencing data.\n",
      "--   Nat Methods. 2013 Jun;10(6):563-9\n",
      "--   http://doi.org/10.1038/nmeth.2474\n",
      "-- \n",
      "-- CONFIGURE CANU\n",
      "--\n",
      "-- Detected Java(TM) Runtime Environment '1.8.0_162' (from 'java').\n",
      "-- Detected gnuplot version '4.6 patchlevel 6' (from 'gnuplot') and image format 'png'.\n",
      "-- Detected 2 CPUs and 4 gigabytes of memory.\n",
      "-- No grid engine detected, grid disabled.\n",
      "--\n",
      "--                            (tag)Concurrency\n",
      "--                     (tag)Threads          |\n",
      "--            (tag)Memory         |          |\n",
      "--        (tag)         |         |          |     total usage     algorithm\n",
      "--        -------  ------  --------   --------  -----------------  -----------------------------\n",
      "-- Local: meryl      4 GB    2 CPUs x   1 job      4 GB    2 CPUs  (k-mer counting)\n",
      "-- Local: cormhap    4 GB    2 CPUs x   1 job      4 GB    2 CPUs  (overlap detection with mhap)\n",
      "-- Local: obtovl     4 GB    2 CPUs x   1 job      4 GB    2 CPUs  (overlap detection)\n",
      "-- Local: utgovl     4 GB    2 CPUs x   1 job      4 GB    2 CPUs  (overlap detection)\n",
      "-- Local: ovb        4 GB    1 CPU  x   1 job      4 GB    1 CPU   (overlap store bucketizer)\n",
      "-- Local: ovs        4 GB    1 CPU  x   1 job      4 GB    1 CPU   (overlap store sorting)\n",
      "-- Local: red        4 GB    2 CPUs x   1 job      4 GB    2 CPUs  (read error detection)\n",
      "-- Local: oea        4 GB    1 CPU  x   1 job      4 GB    1 CPU   (overlap error adjustment)\n",
      "-- Local: bat        4 GB    2 CPUs x   1 job      4 GB    2 CPUs  (contig construction with bogart)\n",
      "-- Local: gfa        4 GB    2 CPUs x   1 job      4 GB    2 CPUs  (GFA alignment and processing)\n",
      "--\n",
      "-- Found Nanopore uncorrected reads in the input files.\n",
      "--\n",
      "-- Generating assembly 'agalactiae' in '/home/jovyan/notebooks/data/agalactiae/canu_output'\n",
      "--\n",
      "-- Parameters:\n",
      "--\n",
      "--  genomeSize        4600000\n",
      "--\n",
      "--  Overlap Generation Limits:\n",
      "--    corOvlErrorRate 0.3200 ( 32.00%)\n",
      "--    obtOvlErrorRate 0.1440 ( 14.40%)\n",
      "--    utgOvlErrorRate 0.1440 ( 14.40%)\n",
      "--\n",
      "--  Overlap Processing Limits:\n",
      "--    corErrorRate    0.5000 ( 50.00%)\n",
      "--    obtErrorRate    0.1440 ( 14.40%)\n",
      "--    utgErrorRate    0.1440 ( 14.40%)\n",
      "--    cnsErrorRate    0.1920 ( 19.20%)\n",
      "--\n",
      "--\n",
      "-- BEGIN CORRECTION\n",
      "--\n",
      "----------------------------------------\n",
      "-- Starting command on Fri Apr 27 15:17:24 2018 with 38.086 GB free disk space\n",
      "\n",
      "    cd .\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/sqStoreCreate \\\n",
      "      -o ./agalactiae.seqStore.BUILDING \\\n",
      "      -minlength 50 \\\n",
      "      ./agalactiae.seqStore.ssi \\\n",
      "    > ./agalactiae.seqStore.BUILDING.err 2>&1\n",
      "\n",
      "-- Finished on Fri Apr 27 15:17:25 2018 (one second) with 38.077 GB free disk space\n",
      "----------------------------------------\n",
      "--\n",
      "-- In sequence store './agalactiae.seqStore':\n",
      "--   Found 3604 reads.\n",
      "--   Found 34840031 bases (7.57 times coverage).\n",
      "--\n",
      "--   Read length histogram (one '*' equals 30.55 reads):\n",
      "--        0   4999   2139 **********************************************************************\n",
      "--     5000   9999    616 ********************\n",
      "--    10000  14999    283 *********\n",
      "--    15000  19999    164 *****\n",
      "--    20000  24999    101 ***\n",
      "--    25000  29999     66 **\n",
      "--    30000  34999     54 *\n",
      "--    35000  39999     29 \n",
      "--    40000  44999     26 \n",
      "--    45000  49999     18 \n",
      "--    50000  54999     13 \n",
      "--    55000  59999      7 \n",
      "--    60000  64999     15 \n",
      "--    65000  69999      6 \n",
      "--    70000  74999      7 \n",
      "--    75000  79999      2 \n",
      "--    80000  84999      6 \n",
      "--    85000  89999      5 \n",
      "--    90000  94999      4 \n",
      "--    95000  99999      3 \n",
      "--   100000 104999      1 \n",
      "--   105000 109999      4 \n",
      "--   110000 114999      3 \n",
      "--   115000 119999      2 \n",
      "--   120000 124999      2 \n",
      "--   125000 129999      1 \n",
      "--   130000 134999      2 \n",
      "--   135000 139999      1 \n",
      "--   140000 144999      2 \n",
      "--   145000 149999      1 \n",
      "--   150000 154999      1 \n",
      "--   155000 159999      3 \n",
      "--   160000 164999      1 \n",
      "--   165000 169999      0 \n",
      "--   170000 174999      0 \n",
      "--   175000 179999      3 \n",
      "--   180000 184999      1 \n",
      "--   185000 189999      2 \n",
      "--   190000 194999      2 \n",
      "--   195000 199999      0 \n",
      "--   200000 204999      1 \n",
      "--   205000 209999      0 \n",
      "--   210000 214999      0 \n",
      "--   215000 219999      0 \n",
      "--   220000 224999      0 \n",
      "--   225000 229999      0 \n",
      "--   230000 234999      0 \n",
      "--   235000 239999      1 \n",
      "--   240000 244999      0 \n",
      "--   245000 249999      0 \n",
      "--   250000 254999      0 \n",
      "--   255000 259999      1 \n",
      "--   260000 264999      0 \n",
      "--   265000 269999      0 \n",
      "--   270000 274999      0 \n",
      "--   275000 279999      1 \n",
      "--   280000 284999      0 \n",
      "--   285000 289999      1 \n",
      "--   290000 294999      0 \n",
      "--   295000 299999      0 \n",
      "--   300000 304999      0 \n",
      "--   305000 309999      0 \n",
      "--   310000 314999      0 \n",
      "--   315000 319999      0 \n",
      "--   320000 324999      0 \n",
      "--   325000 329999      0 \n",
      "--   330000 334999      0 \n",
      "--   335000 339999      0 \n",
      "--   340000 344999      0 \n",
      "--   345000 349999      1 \n",
      "--   350000 354999      0 \n",
      "--   355000 359999      0 \n",
      "--   360000 364999      0 \n",
      "--   365000 369999      0 \n",
      "--   370000 374999      0 \n",
      "--   375000 379999      1 \n",
      "--   380000 384999      0 \n",
      "--   385000 389999      1 \n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'meryl' concurrent execution on Fri Apr 27 15:17:25 2018 with 38.077 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/0-mercounts\n",
      "    ./meryl.sh 1 > ./meryl.000001.out 2>&1\n",
      "\n",
      "-- Finished on Fri Apr 27 15:17:36 2018 (11 seconds) with 37.956 GB free disk space\n",
      "----------------------------------------\n",
      "-- Meryl finished successfully.\n",
      "--\n",
      "--  16-mers                                                                                           Fraction\n",
      "--    Occurrences   NumMers                                                                         Unique Total\n",
      "--       1-     1  28645526 *******************************************************************--> 0.9267 0.8235\n",
      "--       2-     2   1748994 ********************************************************************** 0.9833 0.9240\n",
      "--       3-     4    393824 ***************                                                        0.9933 0.9507\n",
      "--       5-     7     60200 **                                                                     0.9971 0.9651\n",
      "--       8-    11     25634 *                                                                      0.9983 0.9721\n",
      "--      12-    16     19377                                                                        0.9990 0.9786\n",
      "--      17-    22     11297                                                                        0.9996 0.9862\n",
      "--      23-    29      2275                                                                        0.9999 0.9913\n",
      "--      30-    37       719                                                                        0.9999 0.9926\n",
      "--      38-    46       450                                                                        0.9999 0.9933\n",
      "--      47-    56       280                                                                        1.0000 0.9938\n",
      "--      57-    67       227                                                                        1.0000 0.9942\n",
      "--      68-    79       160                                                                        1.0000 0.9946\n",
      "--      80-    92       147                                                                        1.0000 0.9949\n",
      "--      93-   106       106                                                                        1.0000 0.9953\n",
      "--     107-   121        87                                                                        1.0000 0.9956\n",
      "--     122-   137        66                                                                        1.0000 0.9958\n",
      "--     138-   154        69                                                                        1.0000 0.9961\n",
      "--     155-   172        43                                                                        1.0000 0.9964\n",
      "--     173-   191        24                                                                        1.0000 0.9966\n",
      "--     192-   211        20                                                                        1.0000 0.9967\n",
      "--     212-   232        20                                                                        1.0000 0.9968\n",
      "--     233-   254         7                                                                        1.0000 0.9969\n",
      "--     255-   277         7                                                                        1.0000 0.9970\n",
      "--     278-   301         7                                                                        1.0000 0.9971\n",
      "--     302-   326        14                                                                        1.0000 0.9971\n",
      "--     327-   352        14                                                                        1.0000 0.9972\n",
      "--     353-   379        11                                                                        1.0000 0.9974\n",
      "--     380-   407         8                                                                        1.0000 0.9975\n",
      "--     408-   436         9                                                                        1.0000 0.9976\n",
      "--     437-   466         9                                                                        1.0000 0.9977\n",
      "--     467-   497         2                                                                        1.0000 0.9978\n",
      "--     498-   529        12                                                                        1.0000 0.9978\n",
      "--     530-   562         5                                                                        1.0000 0.9980\n",
      "--     563-   596         7                                                                        1.0000 0.9981\n",
      "--     597-   631         5                                                                        1.0000 0.9982\n",
      "--     632-   667         5                                                                        1.0000 0.9983\n",
      "--     668-   704         2                                                                        1.0000 0.9984\n",
      "--     705-   742         1                                                                        1.0000 0.9984\n",
      "--     743-   781         3                                                                        1.0000 0.9985\n",
      "--     782-   821         1                                                                        1.0000 0.9985\n",
      "--\n",
      "--        7424 (max occurrences)\n",
      "--     6140445 (total mers, non-unique)\n",
      "--     2264160 (distinct mers, non-unique)\n",
      "--    28645526 (unique mers)\n",
      "-- For mhap overlapping, set repeat k-mer threshold to 347.\n",
      "--\n",
      "-- Found 34785971 16-mers; 30909686 distinct and 28645526 unique.  Largest count 7424.\n",
      "--\n",
      "-- OVERLAPPER (mhap) (correction)\n",
      "--\n",
      "-- Set corMhapSensitivity=high based on read coverage of 7.\n",
      "--\n",
      "-- PARAMETERS: hashes=768, minMatches=2, threshold=0.78\n",
      "--\n",
      "-- Given 4 GB, can fit 6000 reads per block.\n",
      "-- For 2 blocks, set stride to 2 blocks.\n",
      "-- Logging partitioning to 'correction/1-overlapper/partitioning.log'.\n",
      "-- Configured 1 mhap precompute jobs.\n",
      "-- Configured 1 mhap overlap jobs.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'cormhap' concurrent execution on Fri Apr 27 15:17:37 2018 with 38.077 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/1-overlapper\n",
      "    ./precompute.sh 1 > ./precompute.000001.out 2>&1\n",
      "\n",
      "-- Finished on Fri Apr 27 15:30:15 2018 (758 seconds) with 37.95 GB free disk space\n",
      "----------------------------------------\n",
      "-- All 1 mhap precompute jobs finished successfully.\n",
      "--\n",
      "-- Running jobs.  First attempt out of 2.\n",
      "----------------------------------------\n",
      "-- Starting 'cormhap' concurrent execution on Fri Apr 27 15:30:15 2018 with 37.95 GB free disk space (1 processes; 1 concurrently)\n",
      "\n",
      "    cd correction/1-overlapper\n",
      "    ./mhap.sh 1 > ./mhap.000001.out 2>&1\n",
      "\n",
      "-- Finished on Fri Apr 27 15:30:39 2018 (24 seconds) with 37.947 GB free disk space\n",
      "----------------------------------------\n",
      "-- Found 1 mhap overlap output files.\n",
      "----------------------------------------\n",
      "-- Starting command on Fri Apr 27 15:30:39 2018 with 37.947 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreConfig \\\n",
      "     -S ./agalactiae.seqStore \\\n",
      "     -M 4-4 \\\n",
      "     -L ./1-overlapper/ovljob.files \\\n",
      "     -create ./agalactiae.ovlStore.config \\\n",
      "     > ./agalactiae.ovlStore.config.txt \\\n",
      "    2> ./agalactiae.ovlStore.config.err\n",
      "\n",
      "-- Finished on Fri Apr 27 15:30:39 2018 (lickety-split) with 37.947 GB free disk space\n",
      "----------------------------------------\n",
      "--\n",
      "-- Creating overlap store correction/agalactiae.ovlStore using:\n",
      "--      1 bucket\n",
      "--      1 slice\n",
      "--        using at most 1 GB memory each\n",
      "----------------------------------------\n",
      "-- Starting command on Fri Apr 27 15:30:39 2018 with 37.947 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreBuild \\\n",
      "     -O ./agalactiae.ovlStore.BUILDING \\\n",
      "     -S ./agalactiae.seqStore \\\n",
      "     -C ./agalactiae.ovlStore.config \\\n",
      "     > ./agalactiae.ovlStore.err 2>&1\n",
      "\n",
      "-- Finished on Fri Apr 27 15:30:40 2018 (one second) with 37.923 GB free disk space\n",
      "----------------------------------------\n",
      "-- Checking store.\n",
      "----------------------------------------\n",
      "-- Starting command on Fri Apr 27 15:30:40 2018 with 37.923 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/ovStoreDump \\\n",
      "     -S ./agalactiae.seqStore \\\n",
      "     -O ./agalactiae.ovlStore \\\n",
      "     -counts \\\n",
      "     > ./agalactiae.ovlStore/counts.dat 2> ./agalactiae.ovlStore/counts.err\n",
      "\n",
      "-- Finished on Fri Apr 27 15:30:40 2018 (fuirously fast) with 37.923 GB free disk space\n",
      "----------------------------------------\n",
      "--\n",
      "-- Overlap store 'correction/agalactiae.ovlStore' successfully constructed.\n",
      "-- Found 375014 overlaps for 2360 reads; 1244 reads have no overlaps.\n",
      "--\n",
      "--\n",
      "-- Purged 0.091 GB in 3 overlap output files.\n",
      "-- Set corMinCoverage=0 based on read coverage of 7.\n",
      "-- Global filter scores will be estimated.\n",
      "-- Computing correction layouts.\n",
      "----------------------------------------\n",
      "-- Starting command on Fri Apr 27 15:30:40 2018 with 38.014 GB free disk space\n",
      "\n",
      "    cd correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/generateCorrectionLayouts \\\n",
      "      -S ./agalactiae.seqStore \\\n",
      "      -O ./agalactiae.ovlStore \\\n",
      "      -C ./agalactiae.corStore.WORKING \\\n",
      "      -ec 0 \\\n",
      "      -eC 80 \\\n",
      "    > ./agalactiae.corStore.err 2>&1\n",
      "\n",
      "-- Finished on Fri Apr 27 15:30:40 2018 (in the blink of an eye) with 38.012 GB free disk space\n",
      "----------------------------------------\n",
      "-- Computing correction layouts.\n",
      "----------------------------------------\n",
      "-- Starting command on Fri Apr 27 15:30:40 2018 with 38.012 GB free disk space\n",
      "\n",
      "    cd correction/2-correction\n",
      "    /home/jovyan/software/canu/Linux-amd64/bin/filterCorrectionLayouts \\\n",
      "      -S  ../agalactiae.seqStore \\\n",
      "      -C  ../agalactiae.corStore \\\n",
      "      -R  ./agalactiae.readsToCorrect.WORKING \\\n",
      "      -cc 0 \\\n",
      "      -cl 50 \\\n",
      "      -g  4600000 \\\n",
      "      -c  40 \\\n",
      "    > ./agalactiae.readsToCorrect.err 2>&1\n",
      "\n",
      "-- Finished on Fri Apr 27 15:30:40 2018 (lickety-split) with 38.011 GB free disk space\n",
      "----------------------------------------\n",
      "--                             original      original\n",
      "--                            raw reads     raw reads\n",
      "--   category                w/overlaps  w/o/overlaps\n",
      "--   -------------------- ------------- -------------\n",
      "--   Number of Reads               2322          1282\n",
      "--   Number of Bases           18973958         64708\n",
      "--   Coverage                     4.125         0.014\n",
      "--   Median                        2277             0\n",
      "--   Mean                          8171            50\n",
      "--   N50                          30977          3090\n",
      "--   Minimum                         50             0\n",
      "--   Maximum                     388351          8714\n",
      "--   \n",
      "--                                        --------corrected---------  ----------rescued----------\n",
      "--                             evidence                     expected                     expected\n",
      "--   category                     reads            raw     corrected            raw     corrected\n",
      "--   -------------------- -------------  ------------- -------------  ------------- -------------\n",
      "--   Number of Reads               2360           2321          2321              0             0\n",
      "--   Number of Bases           19038666       18596559      16469975              0             0\n",
      "--   Coverage                     4.139          4.043         3.580          0.000         0.000\n",
      "--   Median                        2257           2277          1520              0             0\n",
      "--   Mean                          8067           8012          7096              0             0\n",
      "--   N50                          30899          29505         32724              0             0\n",
      "--   Minimum                         50             50             4              0             0\n",
      "--   Maximum                     388351         388351        319526              0             0\n",
      "--   \n",
      "--                        --------uncorrected--------\n",
      "--                                           expected\n",
      "--   category                       raw     corrected\n",
      "--   -------------------- ------------- -------------\n",
      "--   Number of Reads               1283          1283\n",
      "--   Number of Bases             442107        320859\n",
      "--   Coverage                     0.096         0.070\n",
      "--   Median                           0             0\n",
      "--   Mean                           344           250\n",
      "--   N50                              0             0\n",
      "--   Minimum                          0             0\n",
      "--   Maximum                     377399        320859\n",
      "--   \n",
      "--   Maximum Memory          4028799956\n",
      "--\n",
      "-- ERROR\n",
      "-- ERROR\n",
      "-- ERROR  Found 1 machine configuration:\n",
      "-- ERROR    class0 - 1 machines with 2 cores with 4 GB memory each.\n",
      "-- ERROR\n",
      "-- ERROR  Task cor can't run on any available machines.\n",
      "-- ERROR  It is requesting:\n",
      "-- ERROR    corMemory=15 memory (gigabytes)\n",
      "-- ERROR    corThreads=4 threads\n",
      "-- ERROR\n",
      "-- ERROR  No available machine configuration can run this task.\n",
      "-- ERROR\n",
      "-- ERROR  Possible solutions:\n",
      "-- ERROR    Change corMemory and/or corThreads\n",
      "-- ERROR\n",
      "\n",
      "ABORT:\n",
      "ABORT: Canu snapshot v1.7 +137 changes (r8829 73d5caa1b1087b65f7853ecbebc1bb1dcbd1bc14)\n",
      "ABORT: Don't panic, but a mostly harmless error occurred and Canu stopped.\n",
      "ABORT: Try restarting.  If that doesn't work, ask for help.\n",
      "ABORT:\n",
      "ABORT:   task cor failed to find a configuration to run on.\n",
      "ABORT:\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "execution_count": 1,
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "canu -p agalactiae \\\n",
    "     -d data/agalactiae/canu_output \\\n",
    "     genomeSize=4.6m \\\n",
    "     useGrid=false \\\n",
    "     minReadLength=50 \\\n",
    "     minOverlapLength=50 \\\n",
    "     -nanopore-raw data/agalactiae/merged-output.fasta\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Racon\n",
    "\n",
    "[Racon](https://github.com/isovic/racon) is a consensus module to correct raw contigs generated by rapid assembly methods which do not include a consensus step. We will take our Canu result in FASTA and the raw reads in FASTQ and generate a new FASTA contig file.\n",
    "\n",
    "Before running Racon, we must align the Canu output to the raw reads file and take the overlaps file in PAF as a parameter for Racon command. This can be done using for example minimap, a rapid aligner for ONT reads.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "minimap data/agalactiae/canu_output/agalactiae.contigs.fasta \\\n",
    "        data/agalactiae/merged-output.fastq \\\n",
    "        > agalactiae.paf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Racon parameters:\n",
    "- -t: Number of threads\n",
    "- Raw reads (FASTQ).\n",
    "- Overlaps in (PAF).\n",
    "- Canu output (FASTA)\n",
    "- Racon output file name (FASTA)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "racon -t 48 \\\n",
    "    data/agalactiae/output/merged-output.fastq \\ \n",
    "    agalactiae.paf \\ \n",
    "    data/agalactiae/canu/agalactiae.contigs.fasta \\ \n",
    "    data/agalactiae_racon.contigs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pilon (Requires Illumina reads)\n",
    "\n",
    "[Pilon](https:github.com/broadinstitute/pilon) is a tool that can be used to improve draft assembly as our case and find variation among strains. It Requires as input a FASTA file and a BAM file of reads aligned to the input FASTA file. At this point, we have the racon contigs file and our reads file (ONT and Illumina reads). We will use the racon contigs file and a BAM file produced by de alignment of ILlumina reads against that contigs file to generate our final result with Pilon.\n",
    "\n",
    "In first place, we need to align our Illumina reads against the racon contigs file. We are using bwa-mem so first we have to index the reference file (racon contigs) and then perform the alignment of the Illumina reads. In order to have the BAM file required by Pilon, we are using samtools to convert bwa output in FASTA to the BAM format.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bwa index data/agalactaie_racon.contigs.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Illumina data not included in repository\n",
    "bwa mem -t 48 \\\n",
    "        data/agalactaie_racon.contigs.fasta  \\\n",
    "        data/Data_Ilumina/Raw/WGS_bacterialIsolates_MiSeq_training-33608589/Sagalactiae_HRC-41106565/Sagal_S5_L001_R1_001.fastq.gz \\\n",
    "        data/Data_Ilumina/Raw/WGS_bacterialIsolates_MiSeq_training-33608589/Sagalactiae_HRC-41106565/Sagal_S5_L001_R2_001.fastq.gz \\\n",
    "        | samtools view -S -b -u - | samtools sort - data/bwa_aligned_reads\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally we index the BAM file an run the Pilon jar:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "samtools index data/bwa_aligned_reads.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pilon arguments:\n",
    "- --threads: Number of threads\n",
    "- --genome: The FASTA input file (racon contigs)\n",
    "- --bam: BAM file (generated by bwa and samtools)\n",
    "- --outdir and --output: Output directory and filename "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "java -Xmx128g -XX:+UseConcMarkSweepGC \\ \n",
    "      -XX:-UseGCOverheadLimit \\ \n",
    "      -jar /home/jovyan/software/pilon/pilon-1.22.jar \\ \n",
    "      --threads 2 \\ \n",
    "      --genome data/agalactaie_racon.contigs.fasta \\ \n",
    "      --bam data/bwa_aligned_reads.bam \\ \n",
    "      --outdir data/agalactiae/pilon_output \\ \n",
    "      --output pilon.contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
