### NanoDJ notebooks index

| Notebook                                | Description                                                  |
| --------------------------------------- | :----------------------------------------------------------- |
| **0_QC_with_Biopython.ipynb**           | Quality control and sequence handling with Biopython.        |
| **1.0_Basecalling_with_Albacore.ipynb** | Getting FASTA files with Albacore and raw FAST5 files.       |
| **1.1_porechop-python.ipynb**           | Sequence trimming and demultiplexing with Porechop.          |
| **2_Assembly_canu.ipynb**               | Assembly pipeline using Canu, Racon and Pilon.               |
| **3_Canu+Nanopolish.ipynb**             | Using Nanopolish modules to improve a Canu assembly.         |
| **4_Flye.ipynb**                        | De novo assembly with Flye software                          |
| **5_Hybrid_assemblies.ipynb**           | Using Nanopore reads along with Illumina reads to produce hybrid assemblies |
| **6_QUAST.ipynb**                       | Assembly comparison in the browser using QUAST               |
| **7_Nanosim.ipynb**                     | Simmulating reads with Nanosim software and Nanosim-h fork with precomputed models |
| **8_Rebaler_alignment.ipynb**           | Reference based assembly using Rebaler                       |
| **9_Bandage.ipynb**                     | Assembly graph Visualization                                 |
| **edu_notebook.ipynb**                  | Quality control and metagenomics exercise using Python and BLAST |

